git-webkit detects a previously-used branch name if the new branch is a substring of an existing branch

ddkilzer opened the pull request:
#### 8d467eedab16185364b783c1157671a877d11775
<pre>
git-webkit detects a previously-used branch name if the new branch is a substring of an existing branch
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=300561">https://bugs.webkit.org/show_bug.cgi?id=300561</a>&gt;
&lt;<a href="https://rdar.apple.com/162446508">rdar://162446508</a>&gt;

Reviewed by Brianna Fan.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
(GitHub.graphql):
- Update mock method to behave like GitHub based on emperical behavior.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.find_existing_pull_request):
- Fix the bug by preventing assignment to `existing_pr` unless the
  branch name actually matches.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
(TestDoPullRequest.test_github_substring_branch): Add.
- Add test that reproduces the behavior after the mock code was updated.

Canonical link: <a href="https://commits.webkit.org/305338@main">https://commits.webkit.org/305338@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7720387e72e6481918e4aeed61ef15dd63b0ab6d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136750 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9110 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48037 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144481 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89723 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/504b3320-46ad-4bd2-984c-df6dcfb10dcf) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138622 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9812 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8956 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104571 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76347 "Build was cancelled. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Skipped layout-tests; layout-tests (exception)") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/d209ceb8-d64e-4489-b53b-8c0ac547a279) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139695 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7174 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122529 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85409 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ce8e5e21-4d6f-4a72-831b-b85e8a182f5f) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/136098 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6818 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4504 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5070 "Built successfully") | | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116139 "Found 1 new API test failure: TestWebKitAPI.WebKit2.RTCDataChannelPostMessage (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40714 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147238 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8793 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41287 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112927 "Passed tests") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/136176 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8811 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7394 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113256 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6737 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118817 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62947 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21268 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8841 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36871 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8562 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8781 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8633 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
```diff
@@ -720,6 +720,67 @@ def test_github_reopen(self):
+            remotes=dict(fork='https://{}/Contributor/WebKit'.format(remote.hosts[0])),
+        ) as repo, mocks.local.Svn(), patch('webkitbugspy.Tracker._trackers', []):
+            # First create a PR with the longer branch name
+            with OutputCapture():
```
There shouldn't be any output to even capture by using `repo.commits` (postulating without checking the code). Anyhow, it'd be better to move this up into the `with` statement above and delete the second one at line 756.

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
```diff
@@ -720,6 +720,67 @@ def test_github_reopen(self):
+
+            # Now create a new branch with a name that is a substring of the closed PR's branch
+            # WITHOUT the fix, this will incorrectly match the closed PR
+            with OutputCapture(level=logging.INFO) as captured, MockTerminal.input('y'):
```
As mentioned above, let's move these statements up into the line 727.

briannafan commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
```diff
@@ -720,6 +720,67 @@ def test_github_reopen(self):
+
+        # Verify that no error message about "already associated" was shown
+        # This is the bug: the old code incorrectly matches closed PRs with substring branch names
+        self.assertNotIn('is already associated with', captured.stdout.getvalue())
```
Now we can replace these lines with the assertion patterns found in `test_github_bugzilla` to verify more granularly that we have the correct behaviour.

briannafan reviewed (approved):
LGTM, thanks!

Commit:
git-webkit detects a previously-used branch name if the new branch is a substring of an existing branch
<https://bugs.webkit.org/show_bug.cgi?id=300561>
<rdar://162446508>

Reviewed by Brianna Fan.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/mocks/remote/git_hub.py:
(GitHub.graphql):
- Update mock method to behave like GitHub based on emperical behavior.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/pull_request.py:
(PullRequest.find_existing_pull_request):
- Fix the bug by preventing assignment to `existing_pr` unless the
  branch name actually matches.
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/pull_request_unittest.py:
(TestDoPullRequest.test_github_substring_branch): Add.
- Add test that reproduces the behavior after the mock code was updated.

Canonical link: https://commits.webkit.org/305338@main
